### ISSUE TITLE
error: StreamTooLong when recompiling; duplicate source files in cache manifest

### ISSUE THREAD

**Issue by rupurt** [ziglang/zig#16149]
### Zig Version

0.11.0-dev.3771+128fd7dd0

### Steps to Reproduce and Observed Behavior

https://github.com/rupurt/record-duckdb-extension

## Steps to reproduce

```
nix develop -c $SHELL
nix run .#build-fast
nix run .#build-fast
```

## Observed behavior

```
build.zig options:
- openssl-include-dir=/nix/store/4a0x3z29dhwid9w3wpc1vl05ykm5pqgs-openssl-3.0.9-dev/include
zig build-lib duckdb_static ReleaseFast native: error: error: StreamTooLong

zig build-lib duckdb_static ReleaseFast native: error: the following command exited with error code 1:
/nix/store/213vknkrkc5izl27z7xzs67rm8m6pvfx-zig-0.11.0-dev.3771+128fd7dd0/bin/zig build-lib @/home/alex/workspace/rupurt/record-duckdb-extension/zig-cache/args/12e63fc7cee3d195fbfe9ad73681e58c0e9a44720a0f0886f19d2df986096ddb
Build Summary: 35/38 steps succeeded; 1 failed (disable with --summary none)
install transitive failure
└─ install duckdb_static transitive failure
   └─ zig build-lib duckdb_static ReleaseFast native failure
error: the following build command failed with exit code 1:
/home/alex/workspace/rupurt/record-duckdb-extension/zig-cache/o/78b16d4b1e14050816deef8c0e0e9dee/build /nix/store/213vknkrkc5izl27z7xzs67rm8m6pvfx-zig-0.11.0-dev.3771+128fd7dd0/bin/zig /home/alex/workspace/rupurt/record-duckdb-extension /home/alex/workspace/rupurt/record-duckdb-extension/zig-cache /home/alex/.cache/zig -Dopenssl-include-dir=/nix/store/4a0x3z29dhwid9w3wpc1vl05ykm5pqgs-openssl-3.0.9-dev/include -Doptimize=ReleaseFast

```

### Expected Behavior

I should be able to compile subsequent runs successfully

**Comment by matu3ba** [comment]
Can you reduce the problem further and provide an ideally minimal reproducible? There are many instances of that error in libstd and your codebase looks bigger:
```
user@pc ~/d/g/z/z/master (master)> rg 'StreamTooLong' lib/std/
lib/std/io/reader.zig
62:        /// `error.StreamTooLong` is returned
85:                    return error.StreamTooLong;
99:        /// memory would be greater than `max_size`, returns `error.StreamTooLong`.
112:        /// If the `std.ArrayList` length would exceed `max_size`, `error.StreamTooLong` is returned and the
126:        /// memory would be greater than `max_size`, returns `error.StreamTooLong`.
143:        /// large enough to hold the entire contents, `error.StreamTooLong` is returned.
159:        /// `error.StreamTooLong`. If end-of-stream is found, returns the rest
183:        /// large enough to hold the entire contents, `error.StreamTooLong` is returned.
206:        /// returns `error.StreamTooLong` and finishes appending.
208:        pub fn streamUntilDelimiter(self: Self, writer: anytype, delimiter: u8, optional_max_size: ?usize) (Error || error{ EndOfStream, StreamTooLong } || @TypeOf(writer).Error)!void {
215:                return error.StreamTooLong;
222:                // Can not throw `error.StreamTooLong` since there are no boundary.
441:test "Reader.readUntilDelimiterArrayList returns StreamTooLong, then an ArrayList with bytes read until the delimiter" {
449:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiterArrayList(&list, '\n', 5));
501:test "Reader.readUntilDelimiterAlloc returns StreamTooLong, then an ArrayList with bytes read until the delimiter" {
507:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiterAlloc(a, '\n', 5));
538:test "Reader.readUntilDelimiter returns StreamTooLong, then an empty string" {
542:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiter(&buf, '\n'));
546:test "Reader.readUntilDelimiter returns StreamTooLong, then bytes read until the delimiter" {
550:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiter(&buf, '\n'));
577:test "Reader.readUntilDelimiter returns StreamTooLong, then EndOfStream" {
581:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiter(&buf, '\n'));
591:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiter(&buf, '\n'));
629:test "Reader.readUntilDelimiterOrEofAlloc returns StreamTooLong, then an ArrayList with bytes read until the delimiter" {
635:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiterOrEofAlloc(a, '\n', 5));
657:test "Reader.readUntilDelimiterOrEof returns StreamTooLong, then an empty string" {
661:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiterOrEof(&buf, '\n'));
665:test "Reader.readUntilDelimiterOrEof returns StreamTooLong, then bytes read until the delimiter" {
669:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiterOrEof(&buf, '\n'));
695:test "Reader.readUntilDelimiterOrEof returns StreamTooLong, then bytes read until end-of-stream" {
699:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiterOrEof(&buf, '\n'));
709:    try std.testing.expectError(error.StreamTooLong, reader.readUntilDelimiterOrEof(&buf, '\n'));
729:    try std.testing.expectError(error.StreamTooLong, reader.streamUntilDelimiter(writer, '!', 5));
```

Also, please provide the `zig` command to reproduce the error to rule out nix being the source of the problem.

**Comment by rupurt** [comment]
Full command with same error:

```sh
zig build -Doptimize=ReleaseFast -Dopenssl-include-dir=/nix/store/gx7946h9y6qwcdil01ibf4hvm8vfc823-openssl-3.0.9-dev/include -Dbuild-shell
```

Unfortunately I don't know how to reduce the problem further. The example steps I gave in the repo produces the error every time. I'm running `zig build ...` twice.

I started looking through the `zig build` implementation and I feel like [this](https://github.com/ziglang/zig/blob/9d66481e3df35b54275373a685e765a4bcebd712/lib/std/Build/Step/Run.zig#L73) could be a good starting point. It seems to hard code a max IO size limit of 10mb. I'm currently trying to figure out how to rebuild zig and use it in this codebase.

**Comment by slimsag** [comment]
I am encountering this frequently when trying to cross-compile mach-gpu-dawn for Windows, so I can at least confirm nix is not the only source of the problem.

Unfortunately I don't have an easy/minimal repro, I'll see if I can produce one later

**Comment by slimsag** [comment]
It is notable that in both my case and @rupurt's report:

* this only appears to happen on the second `zig build`, the first one is fine. If you blow away `zig-cache` and `zig-out`, the issue is worked around until the next build.
* both our failing commands are `zig build-lib` with enough args that a compiler response file `zig build-lib @foo` needs to be used.

**Comment by slimsag** [comment]
OK I've traced it down to the manifest file exceeding 50MB, here's a stack trace from a stage3 zig:

```
/Users/slimsag/Desktop/zig/build/stage3/bin/zig build-lib @/Users/slimsag/Desktop/hexops/mach/libs/gpu/zig-cache/args/6eb4a576449b3bb4f02df34922b46612d50202131eb23e0cbbf92134466cdd62 
zig build-lib dawn ReleaseFast x86_64-windows: error: error: StreamTooLong
/Users/slimsag/Desktop/zig/lib/std/io/reader.zig:85:21: 0x1025bed43 in readAllArrayListAligned__anon_42380 (zig)
                    return error.StreamTooLong;
                    ^
/Users/slimsag/Desktop/zig/lib/std/io/reader.zig:65:13: 0x1025bea03 in readAllArrayList (zig)
            return self.readAllArrayListAligned(null, array_list, max_append_size);
            ^
/Users/slimsag/Desktop/zig/lib/std/io/reader.zig:105:13: 0x1025bef33 in readAllAlloc (zig)
            try self.readAllArrayList(&array_list, max_size);
            ^
/Users/slimsag/Desktop/zig/lib/std/Build/Cache.zig:436:35: 0x1025c124b in hit (zig)
            const file_contents = try self.manifest_file.?.reader().readAllAlloc(gpa, manifest_file_size_max);
                                  ^
/Users/slimsag/Desktop/zig/src/Compilation.zig:1905:52: 0x10265cdcf in update (zig)
            const i = man.failed_file_index orelse return err;
                                                   ^
/Users/slimsag/Desktop/zig/src/main.zig:3420:21: 0x1026876ef in serve (zig)
                    try comp.update(main_progress_node);
                    ^
/Users/slimsag/Desktop/zig/src/main.zig:3234:13: 0x10251ecc3 in buildOutputType (zig)
            try serve(
            ^
/Users/slimsag/Desktop/zig/src/main.zig:271:9: 0x1024d0fdf in mainArgs (zig)
        return buildOutputType(gpa, arena, args, .{ .build = .Lib });
        ^
/Users/slimsag/Desktop/zig/src/main.zig:213:5: 0x1024ce7cf in main (zig)
    return mainArgs(gpa, arena, args);
    ^

zig build-lib dawn ReleaseFast x86_64-windows: error: the following command exited with error code 1:
/Users/slimsag/Desktop/zig/build/stage3/bin/zig build-lib @/Users/slimsag/Desktop/hexops/mach/libs/gpu/zig-cache/args/6eb4a576449b3bb4f02df34922b46612d50202131eb23e0cbbf92134466cdd62 
```

**Comment by slimsag** [comment]
I found the manifest file max is 50MB:

```zig
const manifest_file_size_max = 50 * 1024 * 1024;
```

And began inspecting the manifest files produced for my `mach-gpu-dawn` build, most are <150K, but two are not:

```
122M	./zig-cache/h/326381304e641965cd20afd8120b2761.txt
122M	./zig-cache/h/62e8fa63813e956aff18bd12659f7ffd.txt
```

I don't have a convenient place to upload files that large, but I could if it's helpful. I was able to introspect the contents of them a bit and I think I've found the cause? They are valid manifest files, they look like this:

```
9204 79469445 1687615676000000000 30ba36d2e76cc3f4eb662adecc256507 1 libc/include/any-windows-any/ole2.h
7852 79469433 1687615676000000000 ed27bfcd8ba39c420f3e66f0cf7f8b18 1 libc/include/any-windows-any/objbase.h
14373 79468632 1687615676000000000 e4cfc5527d471f9a0873d486fb40fdb6 1 libc/include/any-windows-any/combaseapi.h
9699 79470009 1687615676000000000 0ead0d629701e46c0acfac8339e10aa4 1 libc/include/any-windows-any/wtypesbase.h
```

There are 1.1 million lines:

```
cat ./zig-cache/h/326381304e641965cd20afd8120b2761.txt | wc -l
 1102133
```

~100k (18MB of the file) are from my project:

```
cat ./zig-cache/h/326381304e641965cd20afd8120b2761.txt | grep '/Users/slimsag/Desktop/hexops/mach' | wc -l
  102061
```

There appear to be millions of repeated, identical lines:

```
slimsag@hexops-ci gpu % cat ./zig-cache/h/326381304e641965cd20afd8120b2761.txt | gsort | wc -l
 1102133
slimsag@hexops-ci gpu % 
slimsag@hexops-ci gpu % cat ./zig-cache/h/326381304e641965cd20afd8120b2761.txt | gsort | uniq | wc -l
    4054
```

If I remove only the 100% duplicate lines, the manifest file is just 704K:

```
slimsag@hexops-ci gpu % cat ./zig-cache/h/326381304e641965cd20afd8120b2761.txt | gsort | uniq > tmpfo
slimsag@hexops-ci gpu % du -sh tmpfo 
704K	tmpfo
```

......soooo, why do manifest files have so many duplicates in them? any guesses @matu3ba ?

**Comment by slimsag** [comment]
The relevant manifest writing code is here: https://github.com/ziglang/zig/blob/88284c124a0d930541f02ae9727118c0724f93f9/lib/std/Build/Cache.zig#L842

I suspect we need some deduplication logic here (or the files list itself should not have duplicates), not 100% sure but it might be a while before I have time to look into this further

**Comment by andrewrk** [comment]
> There are 1.1 million lines:

Goodness!

Um, before adding deduplication logic I'd like to understand why there are duplicates in the first place. Maybe it will reveal another issue.

**Comment by slimsag** [comment]
update: I am still encountering this, both in my project from before and now in a new (also large C++) project.

Observations:

* Using `0.12.0-dev.1751+7fbbeae61`
* It doesn't happen on every `zig build`. Just as before, the first build (when zig-cache and zig-out does not exist) is fine. Subsequent builds are _sometimes_ fine, _sometimes_ not.
* When `StreamTooLong` is observed, I always find a very large txt file in zig-cache:

```
 97M    zig-cache/h/649e59cf152d048f7b98c0d08ff65780.txt
```

Deleting _just_ that one file works around the issue.

Subsequent builds _do not_ always produce a 97M txt file like that, hinting there may be some non-deterministic race condition which leads to the txt file expanding to that size?

**Comment by matu3ba** [comment]
> some non-deterministic race condition which leads to the txt file expanding to that size?

From what I read so far the issue occurs on compiling C or C++ code without needing interaction with Zig code (usage as extern etc + linking).
Is this the behavior consistent across different Kernels (Windows, Linux, Mac?

**Comment by rupurt** [comment]
@matu3ba yes. Both linux x86 + m2 aarch64

**Comment by slimsag** [comment]
Some more observations:

* I encounter this problem infrequently (maybe one out of every 50 or so `zig build`s?)
* I encounter it when building from a macOS M3 machine, for macOS m3, building C++ code using `zig build`.
* When it happens it's always a really large text file was produced somehow:

```
...
164K    zig-cache/h/75d6288b8dd32efb05178b713ec39e84.txt
164K    zig-cache/h/a296413bc05a88a26ff568665eb3a6e1.txt
244K    zig-cache/h/af372330afe7ea1e18b1a10f9f89bc37.txt
 96M    zig-cache/h/e824fea1e6db08475b83b96fa6356a8f.txt
```

I find the text file like this:

```
find zig-cache | grep '\.txt' | xargs du -sh | sort -h
```

Once removed, `zig build` works fine again and then the largest text file found by that command is only 244K, nowhere near the 96M.

I have been reproducing it using this codebase: https://github.com/hexops/mach-dxcompiler/tree/eca3d79b7159306211208a7329e5fd4d7a78d6bc

With this command:

```
zig build -Dfrom-source
```

But it doesn't happen all the time, it's very inconsistent. It seems to happen more frequently when I e.g. change Git branches or do some other action which might invalidate the zig build cache in some larger way.

I have copies of the large text files `zig-cache/h/e824fea1e6db08475b83b96fa6356a8f.txt` etc. (I have saved ~6 of them, one each time I encountered this in the past two weeks) but too large to upload; if you want them DM me on Discord

**Comment by andrewrk** [comment]
I think the next move is to put some extra logic in the compiler to detect this anomalous state earlier and fail sooner. That might lead directly to a solution.

**Comment by hazeycode** [comment]
Encountering 100% of the time on macOS here: https://github.com/zig-gamedev/zig-gamedev/pull/463

**Comment by sergeypdev** [comment]
Got the same issue on macos: https://github.com/sergeypdev/godot-cpp (Zig 0.12.0-dev.2341+92211135f)

**Comment by hazeycode** [comment]
> Encountering 100% of the time on macOS here: [zig-gamedev/zig-gamedev#463](https://github.com/zig-gamedev/zig-gamedev/pull/463) (Zig 0.12.0-dev.2063+804cee3b9)

Also on Linux

**Comment by slimsag** [comment]
I've been able to test with Zig 0.12.0-dev.2665+919a3bae1, i.e. latest Zig nightly as of the time of writing, which should include @jacobly0's change. Unfortunately it seems #18801 didn't fix the issue.

I am now able to reproduce this reliably on 100% of builds (before it was nondeterministic), on a macOS apple silicion host:

```
git clone https://github.com/hexops/mach-dxcompiler
cd mach-dxcompiler
git checkout stream-too-long
zig build -Dfrom-source=true
```

The above will always fail with:

```
install
└─ install machdxcompiler
   └─ zig build-lib machdxcompiler Debug native failure
error: error: StreamTooLong

error: the following command exited with error code 1:
/Users/slimsag/zig/0.12.0-dev.2665+919a3bae1/files/zig build-lib @/Volumes/data/hexops/mach-dxcompiler/zig-cache/args/6423a9fed1e1f0093c628c3fea2697c3f63c344ae4cb0f6c9702f30d047c64b9
```

**Comment by jacobly0** [comment]
* There are `866` source files.
 * Each source file depends on an average of `980` header files.
 * The average manifest line is `121` characters.
 * The whole cache manifest in question is `850775` lines and `103583278` bytes.
 * `866*980 == 848680`
 * `866*980*(121+1) == 103538960`
 * `103583278 > 50*1024*1024`

The math says that we need to either deduplicate headers that are included by multiple files or increase `manifest_file_size_max`.

@slimsag: can you compare `sort manifest.txt | uniq -c | sort -n | tail -1` on the previous manifest files?  I'm getting `857` as the first number for the current reproducible one, which is expected to be less than or equal to the `866` source files.

**Comment by slimsag** [comment]
@jacobly0 that all sounds correct to me and matches what I have seen. My suspicion has always been that the fact there is no deduplication of manifest.txt is the bug. In specific, the repetition of libc entries like this that I had quoted earlier in the issue:

```
9204 79469445 1687615676000000000 30ba36d2e76cc3f4eb662adecc256507 1 libc/include/any-windows-any/ole2.h
7852 79469433 1687615676000000000 ed27bfcd8ba39c420f3e66f0cf7f8b18 1 libc/include/any-windows-any/objbase.h
14373 79468632 1687615676000000000 e4cfc5527d471f9a0873d486fb40fdb6 1 libc/include/any-windows-any/combaseapi.h
9699 79470009 1687615676000000000 0ead0d629701e46c0acfac8339e10aa4 1 libc/include/any-windows-any/wtypesbase.h
```

These were the most repetitive entries and I believe made up the bulk of the file size (though certainly there were some duplicates of non-libc entries, just substantially less.)

I no longer have a good way to reproduce with any other codebase, just [the one I linked last](https://github.com/ziglang/zig/issues/16149#issuecomment-1935335994). What I can confirm for certain with the older codebases where I was able to reproduce before.. is that they were an order of magnitude larger in terms of number of C++ files I was building, and likely amount of headers they were including. I suspect 10-20x larger, as a guess.

The other thing I want to point out is filepath length, since the manifest stores absolute paths, e.g.:

```
4958 3781419 1707372599949978023 3eba3ed9fd9691c435b57065cfbdebb7 0 /Volumes/data/hexops/mach-dxcompiler/libs/DirectXShaderCompiler/tools/clang/lib/Lex/PPConditionalDirectiveRecord.cpp
```

With my other project, filepaths were much deeper and so that likely played a contributing factor.

This also potentially explains some of the non-deterministic behavior I saw in the past (repository checkout at a different location with longer/shorter filepath)

**Comment by andrewrk** [comment]
Noting for future self: when solving this, probably absolute file paths need to be entirely eradicated from this file. I already made *some* progress towards this, for example any zig installation files are stored relative to a prefix. For example, random one from my computer:

```
1332 38670364 1687209192630520142 18432538c3ab3277a7c055dfd1846050 1 libc/musl/include/sys/select.h
```

This is referencing a file that is relative to whatever zig was doing the compilation's lib directory (the `1` indicates this).

So, if we're seeing `/Volumes/data/hexops/mach-dxcompiler/libs/DirectXShaderCompiler/tools/clang/lib/Lex/PPConditionalDirectiveRecord.cpp`, that's an example of this system being bypassed. Ideally, it would be stored relative to the zig package that it lives in.

However, I think this topic that I'm bringing up is a different issue, having more to do with file system robustness and portability, because it's still possible for files to be deeply nested within a package. For example, `libs/DirectXShaderCompiler/tools/clang/lib/Lex/PPConditionalDirectiveRecord.cpp` is still a pretty long string.

When I originally created this system, I did not imagine that there would be 100 MiB of pure file listings, but clearly I did not imagine hard enough. `manifest_file_size_max` is an arbitrary amount. I think it should be bumped 10x, and also the files should be deduplicated, in order to make cache hits faster. It will result in calling fstat() one fewer time for each deduplication.
